### PR TITLE
Dog Lover - Christoforidis Vasilis

### PR DESCRIPTION
**Pull request by chrvaskos**
Hello,

This is my final assignment for the engineering challenge.
I hope you enjoy it as much as I did coding it the past two weeks and I'm looking forward to any questions!

Thank you in advance,
Vasilis

### CODE REVIEW

**Comment by dioannou30** [comment]
Hi Vasilis! 
Thank you very much for your time and effort. I have a few questions if you don't mind answering: 

- Why did you use typescript over plain JS ?
- You used tailwind.css throughout the project. Are you familiar with any other libraries ? why did you choose tailwind over them ?

File: dog-lover/src/components/DogBreedImages/DogBreedImages.test.tsx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import { DefaultRequestBody, rest } from 'msw';
+import { DogBreedImages } from './';
+import { MemoryRouter } from 'react-router-dom';
+
+const server = setupServer(
+  rest.get<DefaultRequestBody>(
+    'https://dog.ceo/api/breed/labrador/images',
+    (_req, res, ctx) => {
+      return res(
+        ctx.delay(100),
```
Comment (kioannou) [inline dog-lover/src/components/DogBreedImages/DogBreedImages.test.tsx:18]: Can you explain why you used delay here ?

File: dog-lover/src/utils/index.tsx [hunk 2]
```diff
@@ -0,0 +1,13 @@
+export const randomItem = <T,>(items: T[]): T => {
+  return items[Math.floor(Math.random() * items.length)];
+};
+
+export const nextItem = <T,>(items: T[], current: T): T => {
```
Comment (kioannou) [inline dog-lover/src/utils/index.tsx:5]: Can you explain why you used generics here? It doesn't seem to be used somewhere

**Review by kioannou — COMMENTED** [review]
Excellent work Vasili !!! Thank you so much for your time mate! can you please check my comments when you have some time? Thanks in advance and again thanks for your effort!

**Comment by chrvaskos** [comment]
> Hi Vasilis! Thank you very much for your time and effort. I have a few questions if you don't mind answering:
> 
> * Why did you use typescript over plain JS ?
> * You used tailwind.css throughout the project. Are you familiar with any other libraries ? why did you choose tailwind over them ?

Hello Dimitris (@dioannou30),

First of all, I want to thank you for your feedback. I really appreciate it!

Moving on to answer your questions:
- Ever since I've started using TypeScript for my personal projects, it has become really hard to look back and use plain JavaScript! TS reduces — or even completely removes — any type errors and makes code refactoring a lot faster. It also acts as a living documentation allowing my IDE to offer reliable auto-completions which eliminate any previous uncertainties or confusions. Honestly, TS forces me to write better code while making the whole process more fun. What's not to like?!
- I personally love Tailwind CSS, because it is flexible and easy to customize while not boxing me into any pre-designed components. I also find it to pair really well with component-based UI libraries/frameworks, such as React in this case, because "HTML bloat" (the one negative I can think of and I've seen being talked about in multiple front-end dev communities/forums) is mitigated to some extent when there are separate reusable components. I'm familiar with lots of different libraries (styled-components, Bootstrap, Material UI to name a few) and I'm pretty confident with pure CSS/SCSS. I've chosen to go with Tailwind because I really enjoy coding with it and I wanted a fast and lightweight solution to build a design from scratch.

Thank you for your time!

File: dog-lover/src/components/DogBreedImages/DogBreedImages.test.tsx (hunk 1 above)
Comment (chrvaskos) [reply dog-lover/src/components/DogBreedImages/DogBreedImages.test.tsx:18]: Hello Kosta,

Thank you so much for taking the time to give me your feedback. It is much appreciated!

In response to your first question, I've only used it to make it a bit more realistic. It is definitely not required!

File: dog-lover/src/utils/index.tsx (hunk 2 above)
Comment (chrvaskos) [reply dog-lover/src/utils/index.tsx:5]: I like to keep any functions in the utils folder as reusable throughout the application as possible. In that way and as the project grows larger, if I need to reuse them with different types of input, I can easily grab and use them right away!

Thanks again for your time!

**Pull request closed without merging**